main.nf

```diff
@@ -495,12 +525,16 @@ if (!params.Q2imported){
 	*/
 	if (!params.multipleSequencingRuns){
 		ch_fastq_trimmed_manifest
-			.map { forward, reverse -> [ forward.drop(forward.findLastIndexOf{"/"})[0], forward, reverse ] } //extract file name
-			.map { name, forward, reverse -> [ name.toString().take(name.toString().indexOf("_")), forward, reverse ] } //extract sample name
-			.map { name, forward, reverse -> [ name +","+ forward + ",forward\n" + name +","+ reverse +",reverse" ] } //prepare basic synthax
+			.map { name, reads ->
+				def sampleID = name 
+				def fwdReads = reads [0]
+				def revReads = reads [1]
+				[ "${sampleID}" +","+ "${fwdReads}" + ",forward\n" + "${sampleID}" +","+ "${revReads}" +",reverse" ]
+			}
```

Comment:
Have to test myself if all that (still) works as expected. Can do that soon-ish.

Reply:
To test whether `--reads` input works as expected I downloaded the test read files into folder data and did:
`nextflow run DiegoBrambilla/ampliseq -profile test,singularity --reads "data" --readPaths false` with the result
```
  ValueError: Sample IDs found in the table are missing in the metadata: {'2a_S115_L001', '2_S115_L001', '1a_S103_L001', '1_S103_L001'}.
  
  Plugin error from taxa:
  
    Sample IDs found in the table are missing in the metadata: {'2a_S115_L001', '2_S115_L001', '1a_S103_L001', '1_S103_L001'}.
  
  See above for debug info.

```
I am pretty certain that comes from your changes in the above mentioned part.

Running `nextflow run DiegoBrambilla/ampliseq -profile test_multi,singularity --reads "data" --readPaths false -resume` all works fine (cause there are no changes).

Reply:
Running
`nextflow run DiegoBrambilla/ampliseq -profile test,singularity --reads "data" --readPaths false`
creates a mismatch between the sample IDs provided by the [test manifest](https://github.com/nf-core/test-datasets/raw/ampliseq/testdata/Metadata_manifest.tsv) and those generated when importing a standard input:
```
Channel
			.fromFilePairs( params.reads + params.extension, size: 2 )
			.ifEmpty { exit 1, "Cannot find any reads matching: ${params.reads}${params.extension}\nNB: Path needs to be enclosed in quotes!" }
			.into { ch_read_pairs; ch_read_pairs_fastqc }
```
The latter creates sample IDs like this: "2a_S115_L001", but downstream analyses expect to receive the same ID as the one reported in the test manifest (something like "2a_S115").
Since this conflict does not verifies when providing raw reads (just tested this weekend), I think updating the [metadata file for the test_manifest](https://github.com/nf-core/test-datasets/raw/ampliseq/testdata/Metadata_manifest.tsv) profile will do the trick.

Reply:
Because I did not specify `--manifest` or `- profile test_manifest`, `test_manifest.config` or `Metadata_manifest.tsv` should have nothing to do with this issue? They simply shouldn't be used as far as I understand. Isn't that so?

Reply:
Right, `Metadata_manifest.tsv` or `test_manifest.config` have nothing to do with this issue.
My bad, I was not clear.
However, the said mismatch between sample IDs generated by `main.nf` and sample IDs generated by [Metadata.tsv](https://github.com/nf-core/test-datasets/raw/ampliseq/testdata/Metadata.tsv) stands.

> They simply shouldn't be used as far as I understand. Isn't that so?
`- profile test_manifest` uses `test_manifest.config` and `Metadata_manifest.tsv`, but you already know this. So yeah, they are not used when calling `-profile test` .

Aside from including `--manifest false` in the command line (the pipeline will have to skip downstream analyses, I guess),  a simple way to solve the issue would be to update sample IDs in [Metadata.tsv](https://github.com/nf-core/test-datasets/raw/ampliseq/testdata/Metadata.tsv) and [params.ReadPaths](https://github.com/DiegoBrambilla/ampliseq/blob/master/conf/test.config#L24-L29) so that they both have sample IDs that also include info about the lane, like "2a_S115_L001".

Reply:
> Aside from including --manifest false in the command line (the pipeline will have to skip downstream analyses, I guess), a simple way to solve the issue would be to update sample IDs in Metadata.tsv and params.ReadPaths so that they both have sample IDs that also include info about the lane, like "2a_S115_L001".

While changing the IDs in Metadata.tsv certainly will make theabove mentioned test work, it is in my opinion not an improvement to use "2a_S115_L001" instead of "2a" as sample ID. The latter seems much more intuitive to me. Isn't there a way where you could keep `--reads` working as is and just add the `--manifest` without breaking existing function? Or would that be overly complicated? Probably I can attempt to help?

Reply:
@d4straub I am grateful for any help you can offer. The present changes to the code are the simplest result I can propose at the best of my current possibilities. 
Since any further improvement of the code would require expertise in groovy that I do not have, I am stuck.
Before discussing the improvements, consider that composite sample IDs are beneficial in some circumstances. For instance, NGI produces raw reads name with this convention:
[POSITION][FLOWCELL]_[NGI-NAME]_[READ]_[LANE].fastq.gz

-     LANE: Sequencing lane that the file originates from
-     POSITION: Position of the flowcell in the sequencer
-     FLOWCELL: Unique flowcell indentifier
-     NGI-NAME: Internal NGI sample indentifier
-     READ: Forward(1) or reverse(2) read indentifier
NGI users detect unique sample IDs by taking "[POSITION][FLOWCELL]_[NGI-NAME]_[READ]" (often you have read pairs that have the same NGI name and differ only by lane or flowcell position). This is actually what started the need for an input manifest file.
About the improvement to the code, I am replicating the same test case you ran and soon will come back with feedback on which part of the code need to be improved.

Reply:
I would like to emphasize that I totally acknowledge the advantage of having a manifest as import option. I just would prefer if it didn't break any existing routines, if possible at all.

Let me know how testing goes.

Reply:
Hi, back with something.
The lines [391-394](https://github.com/DiegoBrambilla/ampliseq/blob/master/main.nf#L391-L394) are the ones that create the sample IDs '2a_S115_L001', '2_S115_L001', '1a_S103_L001', '1_S103_L001' (referring to the aforementioned way you tested ampliseq).
You can verify it by running in your local directory that contains `data` this script:
```
#!/usr/bin/env nextflow

params.reads = "data/*_R{1,2}_001.fastq.gz"

Channel
            .fromFilePairs( params.reads, size: 2 )
            .ifEmpty { exit 1, "Cannot find any reads matching: ${params.reads}${params.extension}\nNB: Path needs to be enclosed in quotes!" }
            .set { ch_read_pairs }
            

process foo {
    echo true
    input:
    set sampleId, file(read) from ch_read_pairs

    script:
    """
    echo tell_me --sample $sampleId --reads $read
    """
}
```
these values are carried until lines [526-536](https://github.com/DiegoBrambilla/ampliseq/blob/master/main.nf#L526-L536) and used without any parsing to build the manifest file for QIIME2 later on in the code. These sample IDs are in contrast with the ones in the test manifest file, as agreed on before.
Ideas on how to solve this:

- Change sample IDs in the [manifest](https://github.com/nf-core/test-datasets/raw/ampliseq/testdata/Metadata.tsv) as follows: '2a_S115_L001', '2_S115_L001', '1a_S103_L001', '1_S103_L001' (deprecated by @d4straub )

- introduce a code snippet [here](https://github.com/DiegoBrambilla/ampliseq/blob/master/main.nf#L391-L394) that parses sample IDs in a smart way, similarly to how it was done [here](https://github.com/nf-core/ampliseq/blob/master/main.nf#L497-L499) in the original ampliseq but without stripping everything comes after the first "_" (0 ideas on how to do it)

- Of course, another possibility would be to put lines [526-536](https://github.com/DiegoBrambilla/ampliseq/blob/master/main.nf#L526-L536) under an if clause ( `if (!params.multipleSequencingRuns && params.manifest) {` ) while keeping lines [497-503](https://github.com/nf-core/ampliseq/blob/master/main.nf#L497-L503) that were present before this PR ( putting them under `} else if (!params.manifest && !params.multipleSequencingRuns ) {`

Tell me what you think or if you have other ideas.

Reply:
Of the three options the last one seems most appealing to me (if clauses) because it is easiest to understand and alter when codes needs to be changed in future.

Could you also add somewhere [here](https://github.com/DiegoBrambilla/ampliseq/blob/02564344b7ef09e67818c8cda0a790bba89737d2/main.nf#L187-L202) a sanity test such as 

```
if (params.multipleSequencingRuns && params.manifest) { 
	exit 1, "The manifest file does not support multiple sequencing runs at this point."
}
```

Reply:
I'm pitching, in realizing I'm not as into the code as the two of you are, so bear with my potential ignorance.

As I understand it, the problem arises from that populating the ch_read_pairs in lines 391-4 does not parse out the sample names from the filenames the way it is when fetching read files from urls (readPaths()). This has the consequence that the trimming process will not pass on pair_ids/sample names to the ch_fastq_trimmed_manifest, which was rewritten by @DiegoBrambilla not to parse file names again. True?

In the interest of managing future code, @DiegoBrambilla's second option, parsing the file names in lines 391-4 sounds like the most robust way, as the code that writes the QIIME manifest would not have to be duplicated. (I'm of course partial here, as I was the one suggesting this option in the first place, again in the interest of not duplicating code but keeping the if clauses limited to what happens first in the script.)

I'm perfectly willing to accept your judgement, guys! :-)